gitignore_tidy/core.py

```diff
@@ -1,123 +1,184 @@
         return
 
-    sorted_lines = _tidy_lines(
-        lines,
-        allow_leading_whitespace=allow_leading_whitespace,
-    )
-
-    _write_if_changed(sorted_lines, lines, path)
-
-
-def _normalize(lines: list[str], allow_leading_whitespace: bool = False) -> list[str]:
-
-    if not allow_leading_whitespace:
-        lines = [re.sub("^(!)? *\t*", "\\1", line) for line in lines]
-    lines = [re.sub(" *\t*$", "", line) for line in lines]
-    unique = []
-
-    for line in lines:
-        if line not in unique or line == "":
-            unique.append(line)
-    return unique
-
-
-def _sort_lines(lines: list[str]) -> list[str]:
-    original = lines
-    non_negated = [re.sub("^!", "", line) for line in lines]
-    sorted_non_negated = sorted(non_negated)
-    sorted_negated = []
-    for line in sorted_non_negated:
-        negated_cand = "!" + line
-        if negated_cand in original:
-            sorted_negated.append(negated_cand)
-        else:
-            sorted_negated.append(line)
-
-    return sorted_negated
-
+    tidy_plain_lines = tidy_lines(lines, allow_leading_whitespace=allow_leading_whitespace)
+    if lines.lines == tidy_plain_lines.lines:
+        logger.info(f"{path} already tidy.")  # TODO use logger module
+    else:
+        tidy_plain_lines.to_file(path)
+        logger.info(f"Successfully written {path}.")
+
+
+def tidy_lines(lines: "PlainLines", allow_leading_whitespace: bool) -> "PlainLines":
+    normalised_contents = lines.normalize(allow_leading_whitespace=allow_leading_whitespace)
+    sorted_sections = Sections(tuple(section.sort() for section in normalised_contents.split()))
+    return sorted_sections.as_plain()
+
+
+@dataclasses.dataclass(frozen=True)
+class PlainLines:
+    """
+    Flat representation of a `.gitignore` file or parts of it
+    """
+
+    lines: list[str]
+
+    normalised: bool = False
+    sorted: bool = False
+
+    @classmethod
+    def from_file(cls, path: pathlib.Path) -> "PlainLines":
+        if not path.exists():
+            raise FileNotFoundError(f"{path} not found.")
+
+        with path.open("r") as f:
+            lines = f.read().splitlines()
+
+        return cls(lines, normalised=False, sorted=False)
+
+    def to_file(self, path: pathlib.Path):
+        with path.open("w") as f:
+            f.writelines([line + "\n" for line in self.lines])
+
+    def normalize(self, allow_leading_whitespace: bool = False) -> "PlainLines":
+        lines = self.lines
+        if not allow_leading_whitespace:
+            lines = [re.sub("^(!)? *\t*", "\\1", line) for line in lines]
+        lines = [re.sub(" *\t*$", "", line) for line in lines]
+        unique = []
+
+        for line in lines:
+            if line not in unique or line == "":
+                unique.append(line)
+        return PlainLines(unique, normalised=True, sorted=self.sorted)
+
+    def __iter__(self) -> typing.Iterator[str]:
+        return iter(self.lines)
+
+    def __len__(self) -> int:
```

Comment:
I am new to this 😀 
> You're exposing parts of the `collections.abc.Sequence` protocol.

Because I call `len()` on a `Sequence`? Is that a problem? I guess I thought I'd just be explicit here. If I understand correctly, subclassing `Sequence` would also mean I'd inherit methods from `Sequence` that I might not need / want?

Reply:
Indeed! This design choice boils down to: fully custom class (status quo) vs. generic class (i.e. Sequence[str] only for the benefit of type checkers) vs. custom subclass e.g. of tuple[str] (or mutable sequence type, depending on other design choices).